mabwiser/approximate.py

```diff
@@ -0,0 +1,170 @@
+                predictions[index] = self._get_no_nhood_predictions(lp, is_predict)
+
+        return predictions
+
+
+class _LSHNearest(_ApproximateNeighbors):
+
+    def __init__(self, rng: _BaseRNG, arms: List[Arm], n_jobs: int, backend: Optional[str],
+                 lp: Union[_EpsilonGreedy, _Linear, _Popularity, _Random, _Softmax, _ThompsonSampling, _UCB1],
+                 n_dimensions: int, n_tables: int, no_nhood_prob_of_arm=Optional[List]):
+        super().__init__(rng, arms, n_jobs, backend, lp, metric='simhash', no_nhood_prob_of_arm=no_nhood_prob_of_arm)
+
+        # Properties for hash tables
+        self.n_dimensions = n_dimensions
+        self.n_tables = n_tables
+        self.buckets = 2 ** n_dimensions
+
+        # Initialize dictionaries for planes and hash table
+        self.table_to_hash_to_index = {k: defaultdict(list) for k in range(self.n_tables)}
+        self.table_to_plane = {i: [] for i in range(self.n_tables)}
+
+    def _add_neighbors(self, hash_values, k, h):
+        self.table_to_hash_to_index[k][h] += list(np.where(hash_values == h)[0])
+
+    def _fit_operation(self, contexts):
+        # Get hashes for each hash table for each training context
+        for k in self.table_to_plane.keys():
+            n_contexts = len(contexts)
+
+            # Partition contexts by job
+            n_jobs, n_contexts, starts = self._partition_contexts(n_contexts)
+
+            # Get hashes in parallel
+            hash_values = Parallel(n_jobs=n_jobs, backend=self.backend)(
+                delayed(self.get_context_hash)(
+                    contexts[starts[i]:starts[i + 1]],
+                    self.table_to_plane[k])
+                for i in range(n_jobs))
+
+            # Reduce
+            hash_values = list(chain.from_iterable(t for t in hash_values))
+
+            # Get list of unique hashes - list is sparse, there should be collisions
+            hash_keys = np.unique(hash_values)
+
+            # For each hash, get the indices of contexts with that hash
+            Parallel(n_jobs=n_jobs, require='sharedmem')(
+                delayed(self._add_neighbors)(
+                    hash_values, k, h)
+                for h in hash_keys)
+
+    def _initialize(self, n_rows):
+        self.table_to_plane = {i: self.rng.standard_normal(size=(n_rows, self.n_dimensions))
+                               for i in self.table_to_plane.keys()}
+
+    def _get_neighbors(self, row_2d):
+        indices = list()
+
+        # Get list of neighbors from each hash table based on the hash values of the new context
+        for k in self.table_to_plane.keys():
+            hash_value = self.get_context_hash(row_2d, self.table_to_plane[k])
+            indices += self.table_to_hash_to_index[k][hash_value[0]]
+
+        return indices
+
+    @staticmethod
+    def get_context_hash(contexts, plane):
+        # Project rows onto plane and get signs
+        projection_signs = 1 * (np.dot(contexts, plane) > 0)
```

Comment:
Does the method make specific assumptions about the mean/variance of the contexts?

Reply:
No, it's just meant to approximate the cosine similarity.